pkg/loghttp/push/otlp.go

```diff
@@ -0,0 +1,322 @@
+
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/grafana/loki/pkg/logql/syntax"
+	"github.com/grafana/loki/pkg/push"
+)
+
+const (
+	pbContentType       = "application/x-protobuf"
+	gzipContentEncoding = "gzip"
+)
+
+var blessedAttributes = []string{
+	"service.name",
+	"service.namespace",
+	"service.instance.id",
+	"deployment.environment",
+	"cloud.region",
+	"cloud.availability_zone",
+	"k8s.cluster.name",
+	"k8s.namespace.name",
+	"k8s.pod.name",
+	"k8s.container.name",
+	"container.name",
+	"k8s.replicaset.name",
+	"k8s.deployment.name",
+	"k8s.statefulset.name",
+	"k8s.daemonset.name",
+	"k8s.cronjob.name",
+	"k8s.job.name",
+}
+
+type pushStats struct {
+	errs                     []error
+	numLines                 int64
+	logLinesBytes            map[time.Duration]int64
+	structuredMetadataBytes  map[time.Duration]int64
+	streamLabelsSize         int64
+	mostRecentEntryTimestamp time.Time
+}
+
+func newPushStats() *pushStats {
+	return &pushStats{
+		logLinesBytes:           map[time.Duration]int64{},
+		structuredMetadataBytes: map[time.Duration]int64{},
+	}
+}
+
+func extractLogs(r *http.Request) (plog.Logs, error) {
+	reader := r.Body
+	if r.Header.Get("Content-Encoding") == gzipContentEncoding {
+		r, err := gzip.NewReader(reader)
+		if err != nil {
+			return plog.NewLogs(), err
+		}
+		reader = r
+		defer func(reader *gzip.Reader) {
+			_ = reader.Close()
+		}(r)
+	}
+	buf, err := io.ReadAll(reader)
+	if err != nil {
+		return plog.NewLogs(), err
+	}
+
+	req := plogotlp.NewExportRequest()
+
+	switch r.Header.Get("Content-Type") {
+	case pbContentType:
+		err := req.UnmarshalProto(buf)
+		if err != nil {
+			return plog.NewLogs(), err
+		}
+	case applicationJSON:
+		err := req.UnmarshalJSON(buf)
+		if err != nil {
+			return plog.NewLogs(), err
+		}
+	default:
+		return plog.NewLogs(),
+			errors.Errorf(
+				"content type: %s is not supported",
+				r.Header.Get("Content-Type"),
+			)
+	}
+
+	return req.Logs(), nil
+}
+
+func otlpToLokiPushRequest(ld plog.Logs, userID string, tenantsRetention TenantsRetention) (*logproto.PushRequest, *pushStats) {
+	stats := newPushStats()
+
+	if ld.LogRecordCount() == 0 {
+		return &logproto.PushRequest{}, stats
+	}
+
+	pushRequestsByStream := map[string]logproto.Stream{}
+
+	rls := ld.ResourceLogs()
+	for i := 0; i < rls.Len(); i++ {
+		sls := rls.At(i).ScopeLogs()
+
+		streamLabels := model.LabelSet{}
+		flattenedResourceAttributes := labels.NewBuilder(logproto.FromLabelAdaptersToLabels(flattenAttributes(rls.At(i).Resource().Attributes(), "")))
+		// service.name is a required Resource Attribute. If it is not present, we will set it to "unknown_service".
+		if flattenedResourceAttributes.Get("service_name") == "" {
+			flattenedResourceAttributes = flattenedResourceAttributes.Set("service_name", "unknown_service")
+		}
+
+		// copy blessed attributes to stream labels
+		for _, ba := range blessedAttributes {
+			normalizedBlessedAttribute := prometheustranslator.NormalizeLabel(ba)
```

Comment:
we do not need to normalize it at runtime, we can generate normalizedBlessedAttributes on startup